ui/src/IUserInterface.ts

```diff
@@ -0,0 +1,32 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { QuickPickItem } from 'vscode';
+
+export interface IUserInterface {
+    showQuickPick<T>(items: PickWithData<T>[] | Thenable<PickWithData<T>[]>, placeHolder: string, ignoreFocusOut?: boolean): Promise<PickWithData<T>>;
+    showQuickPick(items: Pick[] | Thenable<Pick[]>, placeHolder: string, ignoreFocusOut?: boolean): Promise<Pick>;
+
+    showInputBox(placeHolder: string, prompt: string, ignoreFocusOut?: boolean, validateInput?: (s: string) => string | undefined | null, value?: string): Promise<string>;
+
+    showFolderDialog(): Promise<string>;
+}
+
+export class Pick implements QuickPickItem {
+    public readonly description: string;
+    public readonly label: string;
+    constructor(label: string, description?: string) {
+        this.label = label;
+        this.description = description ? description : '';
+    }
+}
+
+export class PickWithData<T> extends Pick {
```

Comment:
How about get rid of PickWithData<T> and use something like:

class Pick<T = void> ...

?

Reply:
Or can you do  

class Pick<T> implement Pick

like you can in C#?

Reply:
Filed an issue to address later:
https://github.com/Microsoft/vscode-azuretools/issues/23